File: src/main/java/com/regionrental/managers/WorldEditManager.java
package com.regionrental.managers;

import com.regionrental.RegionRental;
import com.sk89q.worldedit.EditSession;
import com.sk89q.worldedit.WorldEdit;
import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.extent.clipboard.BlockArrayClipboard;
import com.sk89q.worldedit.extent.clipboard.Clipboard;
import com.sk89q.worldedit.extent.clipboard.io.BuiltInClipboardFormat;
import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormat;
import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormats;
import com.sk89q.worldedit.extent.clipboard.io.ClipboardReader;
import com.sk89q.worldedit.extent.clipboard.io.ClipboardWriter;
import com.sk89q.worldedit.function.operation.ForwardExtentCopy;
import com.sk89q.worldedit.function.operation.Operation;
import com.sk89q.worldedit.function.operation.Operations;
import com.sk89q.worldedit.session.ClipboardHolder;
import com.sk89q.worldedit.math.BlockVector3;
import com.sk89q.worldedit.regions.CuboidRegion;
import com.sk89q.worldedit.regions.Region;
import com.sk89q.worldguard.WorldGuard;
import com.sk89q.worldguard.protection.managers.RegionManager;
import com.sk89q.worldguard.protection.regions.ProtectedRegion;
import org.bukkit.Bukkit;
import org.bukkit.World;

import java.io.*;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Level;

/**
 * Manages WorldEdit integration for block restoration
 * Captures region state when rental starts and restores it when rental expires
 * Uses Sponge Schematic format (.schem) for reliable cross-session persistence
 */
public class WorldEditManager {

    private final RegionRental plugin;
    private final File schematicsFolder;

    // LRU cache for clipboards - limits memory usage
    private static final int DEFAULT_CACHE_SIZE = 20;
    private final int maxCacheSize;
    private final Map<String, Clipboard> clipboardCache;

    // Track which schematics exist on disk (for hasCapture checks without loading)
    // Using thread-safe set for potential concurrent access
    private final Set<String> knownSchematics = ConcurrentHashMap.newKeySet();

    // File extension constants for Sponge schematic format
    private static final String SCHEMATIC_EXTENSION = ".schem";
    private static final String OLD_EXTENSION = ".dat";

    public WorldEditManager(RegionRental plugin) {
        this.plugin = plugin;
        this.schematicsFolder = new File(plugin.getDataFolder(), "schematics");
        this.maxCacheSize = plugin.getConfigManager().getSchematicCacheSize(DEFAULT_CACHE_SIZE);

        // Create LRU cache with access-order ordering, wrapped for thread safety
        this.clipboardCache = Collections.synchronizedMap(new LinkedHashMap<String, Clipboard>(maxCacheSize, 0.75f, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry<String, Clipboard> eldest) {
                boolean shouldRemove = size() > maxCacheSize;
                if (shouldRemove && plugin.getConfigManager().isDebug()) {
                    plugin.getLogger().info("Evicting schematic from cache: " + eldest.getKey());
                }
                return shouldRemove;
            }
        });

        // Create schematics folder if it doesn't exist
        if (!schematicsFolder.exists()) {
            schematicsFolder.mkdirs();
        }

        // Index existing schematics (don't load them into memory)
        indexSchematics();
    }

    /**
     * Captures the current state of a WorldGuard region
     * @param regionName The WorldGuard region name
     * @param world The world the region is in
     * @return true if capture was successful
     */
    public boolean captureRegion(String regionName, World world) {
        if (!plugin.getConfigManager().isBlockRestoration()) {
            return false; // Block restoration disabled
        }

        try {
            // Get the WorldGuard region
            ProtectedRegion wgRegion = getWorldGuardRegion(regionName, world);
            if (wgRegion == null) {
                plugin.getLogger().warning("Could not find WorldGuard region: " + regionName + " in world " + world.getName());
                return false;
            }

            // Convert Bukkit world to WorldEdit world
            com.sk89q.worldedit.world.World weWorld = BukkitAdapter.adapt(world);

            // Convert WorldGuard region to WorldEdit region
            BlockVector3 min = wgRegion.getMinimumPoint();
            BlockVector3 max = wgRegion.getMaximumPoint();
            CuboidRegion region = new CuboidRegion(weWorld, min, max);

            // Create clipboard
            BlockArrayClipboard clipboard = new BlockArrayClipboard(region);

            // Copy blocks to clipboard
            try (EditSession editSession = WorldEdit.getInstance().newEditSession(weWorld)) {
                ForwardExtentCopy copy = new ForwardExtentCopy(
                    editSession, region, clipboard, region.getMinimumPoint()
                );

                // Copy everything including entities
                copy.setCopyingEntities(true);
                copy.setCopyingBiomes(false); // Don't copy biomes for performance

                Operations.complete(copy);
            }

            // Store clipboard with world-aware key
            String compositeKey = world.getName() + ":" + regionName;
            clipboardCache.put(compositeKey, clipboard);
            knownSchematics.add(compositeKey);

            // Save to disk with world-aware filename
            saveSchematic(compositeKey, clipboard);

            if (plugin.getConfigManager().isDebug()) {
                plugin.getLogger().info("Captured region state for: " + regionName + " in world " + world.getName());
            }

            return true;

        } catch (Exception e) {
            plugin.getLogger().log(Level.SEVERE, "Failed to capture region: " + regionName + " in world " + world.getName(), e);
            return false;
        }
    }

    // Backward compatibility - uses first world
    @Deprecated
    public boolean captureRegion(String regionName) {
        World world = Bukkit.getWorlds().get(0);
        return captureRegion(regionName, world);
    }

    /**
     * Restores a region to its captured state
     * @param regionName The WorldGuard region name
     * @param world The world the region is in
     * @return true if restoration was successful
     */
    public boolean restoreRegion(String regionName, World world) {
        if (!plugin.getConfigManager().isBlockRestoration()) {
            return false; // Block restoration disabled
        }

        String compositeKey = world.getName() + ":" + regionName;

        // Lazy load: try cache first, then load from disk
        Clipboard clipboard = getOrLoadClipboard(compositeKey);
        if (clipboard == null) {
            plugin.getLogger().warning("No saved state found for region: " + regionName + " in world " + world.getName());
            return false;
        }

        try {
            // Convert Bukkit world to WorldEdit world
            com.sk89q.worldedit.world.World weWorld = BukkitAdapter.adapt(world);

            // Paste the clipboard back
            try (EditSession editSession = WorldEdit.getInstance().newEditSession(weWorld)) {
                // Use ClipboardHolder with builder pattern for WorldEdit 7.3.16+
                ClipboardHolder holder = new ClipboardHolder(clipboard);
                Operation operation = holder.createPaste(editSession)
                    .to(clipboard.getOrigin())
                    .ignoreAirBlocks(false)
                    .build();
                Operations.complete(operation);
            }

            if (plugin.getConfigManager().isDebug()) {
                plugin.getLogger().info("Restored region state for: " + regionName + " in world " + world.getName());
            }

            // Remove from memory and disk if auto-delete is enabled
            if (plugin.getConfigManager().isAutoDeleteSchematics()) {
                clipboardCache.remove(compositeKey);
                knownSchematics.remove(compositeKey);
                deleteSchematic(compositeKey);
            }

            return true;

        } catch (Exception e) {
            plugin.getLogger().log(Level.SEVERE, "Failed to restore region: " + regionName + " in world " + world.getName(), e);
            return false;
        }
    }

    // Backward compatibility - uses first world
    @Deprecated
    public boolean restoreRegion(String regionName) {
        World world = Bukkit.getWorlds().get(0);
        return restoreRegion(regionName, world);
    }

    /**
     * Checks if a region state has been captured (checks index, doesn't load from disk)
     */
    public boolean hasCapture(String regionName) {
        return knownSchematics.contains(regionName);
    }

    /**
     * Gets the captured clipboard for a region (lazy loads from disk if needed)
     * @param regionName The WorldGuard region name (or composite key world:region)
     * @return The clipboard, or null if not found
     */
    public Clipboard getClipboard(String regionName) {
        return getOrLoadClipboard(regionName);
    }

    /**
     * Deletes a captured region state from cache, index, and disk
     */
    public void deleteCapture(String regionName) {
        clipboardCache.remove(regionName);
        knownSchematics.remove(regionName);
        deleteSchematic(regionName);
    }

    /**
     * Gets the current cache size (for monitoring/debugging)
     */
    public int getCacheSize() {
        return clipboardCache.size();
    }

    /**
     * Gets the total number of known schematics on disk
     */
    public int getTotalSchematicCount() {
        return knownSchematics.size();
    }

    /**
     * Clears the memory cache (schematics remain on disk)
     * Useful for freeing memory if needed
     */
    public void clearCache() {
        clipboardCache.clear();
        if (plugin.getConfigManager().isDebug()) {
            plugin.getLogger().info("Schematic cache cleared");
        }
    }

    /**
     * Lazy loads a clipboard from cache or disk
     * @param compositeKey The composite key (world:region)
     * @return The clipboard, or null if not found
     */
    private Clipboard getOrLoadClipboard(String compositeKey) {
        // Check cache first
        Clipboard cached = clipboardCache.get(compositeKey);
        if (cached != null) {
            return cached;
        }

        // Not in cache - check if it exists on disk
        if (!knownSchematics.contains(compositeKey)) {
            return null;
        }

        // Load from disk
        Clipboard loaded = loadSchematic(compositeKey);
        if (loaded != null) {
            clipboardCache.put(compositeKey, loaded);
            if (plugin.getConfigManager().isDebug()) {
                plugin.getLogger().info("Lazy loaded schematic from disk: " + compositeKey);
            }
        }
        return loaded;
    }

    /**
     * Gets the WorldGuard region
     */
    private ProtectedRegion getWorldGuardRegion(String regionName, World world) {
        RegionManager regionManager = WorldGuard.getInstance()
            .getPlatform()
            .getRegionContainer()
            .get(BukkitAdapter.adapt(world));

        if (regionManager == null) {
            return null;
        }

        return regionManager.getRegion(regionName);
    }

    /**
     * Saves a clipboard to disk using Sponge Schematic format
     * This uses WorldEdit's native format for reliable persistence
     *
     * @param regionName The composite key (world:region)
     * @param clipboard The clipboard to save
     */
    private void saveSchematic(String regionName, Clipboard clipboard) {
        File file = new File(schematicsFolder, regionName + SCHEMATIC_EXTENSION);

        try (FileOutputStream fos = new FileOutputStream(file);
             ClipboardWriter writer = BuiltInClipboardFormat.SPONGE_SCHEMATIC.getWriter(fos)) {
            writer.write(clipboard);
            if (plugin.getConfigManager().isDebug()) {
                plugin.getLogger().info("Saved schematic: " + regionName + SCHEMATIC_EXTENSION);
            }
        } catch (Exception e) {
            plugin.getLogger().log(Level.WARNING, "Failed to save schematic for: " + regionName, e);
        }
    }

    /**
     * Indexes existing schematics without loading them into memory.
     * This allows hasCapture() to work without memory overhead.
     * Also handles migration from legacy .dat files.
     */
    private void indexSchematics() {
        if (!schematicsFolder.exists()) {
            return;
        }

        // Index new .schem files
        File[] schemFiles = schematicsFolder.listFiles((dir, name) -> name.endsWith(SCHEMATIC_EXTENSION));
        if (schemFiles != null) {
            for (File file : schemFiles) {
                String regionName = file.getName().replace(SCHEMATIC_EXTENSION, "");
                knownSchematics.add(regionName);
            }
        }

        // Migration: Handle legacy .dat files
        File[] datFiles = schematicsFolder.listFiles((dir, name) -> name.endsWith(OLD_EXTENSION));
        if (datFiles != null && datFiles.length > 0) {
            plugin.getLogger().warning("Found " + datFiles.length + " legacy .dat file(s) - these are EMPTY due to serialization bug in v2.5.1");
            if (plugin.getConfigManager().isAutoDeleteSchematics()) {
                int deleted = 0;
                for (File datFile : datFiles) {
                    if (datFile.delete()) {
                        deleted++;
                    }
                }
                plugin.getLogger().info("Deleted " + deleted + " empty legacy .dat file(s)");
                if (deleted < datFiles.length) {
                    plugin.getLogger().warning("Could not delete " + (datFiles.length - deleted) + " legacy file(s) - check file permissions");
                }
            } else {
                plugin.getLogger().info("Set 'restoration.auto-delete-schematics: true' to auto-clean legacy files");
            }
        }

        if (!knownSchematics.isEmpty()) {
            plugin.getLogger().info("Indexed " + knownSchematics.size() + " region schematic(s) (lazy loading enabled)");
        }
    }

    /**
     * Loads a schematic from disk using auto-detected format
     * Supports both new .schem format and legacy .dat migration
     *
     * @param regionName The composite key (world:region)
     * @return The loaded clipboard, or null if failed
     */
    private Clipboard loadSchematic(String regionName) {
        // Try new .schem format first
        File schemFile = new File(schematicsFolder, regionName + SCHEMATIC_EXTENSION);

        if (schemFile.exists()) {
            try {
                ClipboardFormat format = ClipboardFormats.findByFile(schemFile);
                if (format == null) {
                    plugin.getLogger().warning("Unknown schematic format: " + schemFile.getName());
                    return null;
                }

                try (FileInputStream fis = new FileInputStream(schemFile);
                     ClipboardReader reader = format.getReader(fis)) {
                    Clipboard clipboard = reader.read();
                    if (plugin.getConfigManager().isDebug()) {
                        plugin.getLogger().info("Loaded schematic: " + regionName + SCHEMATIC_EXTENSION);
                    }
                    return clipboard;
                }
            } catch (Exception e) {
                plugin.getLogger().log(Level.WARNING, "Failed to load schematic: " + schemFile.getName(), e);
                return null;
            }
        }

        // Migration: Warn about legacy .dat files (they're empty due to bug)
        File oldFile = new File(schematicsFolder, regionName + OLD_EXTENSION);
        if (oldFile.exists()) {
            plugin.getLogger().warning("Legacy .dat schematic for '" + regionName + "' is empty (serialization bug). Will be re-captured on next rental.");
            if (plugin.getConfigManager().isAutoDeleteSchematics()) {
                if (!oldFile.delete()) {
                    plugin.getLogger().warning("Could not delete legacy schematic: " + oldFile.getName());
                }
            }
        }

        return null;
    }

    /**
     * Deletes a schematic file from disk
     * Handles both new .schem and legacy .dat formats
     *
     * @param regionName The composite key (world:region)
     */
    private void deleteSchematic(String regionName) {
        // Delete new .schem file
        File schemFile = new File(schematicsFolder, regionName + SCHEMATIC_EXTENSION);
        if (schemFile.exists()) {
            if (!schemFile.delete()) {
                plugin.getLogger().warning("Could not delete schematic: " + schemFile.getName());
            } else if (plugin.getConfigManager().isDebug()) {
                plugin.getLogger().info("Deleted schematic: " + schemFile.getName());
            }
        }

        // Also delete legacy .dat file if it exists (cleanup)
        File oldFile = new File(schematicsFolder, regionName + OLD_EXTENSION);
        if (oldFile.exists()) {
            if (!oldFile.delete()) {
                plugin.getLogger().warning("Could not delete legacy schematic: " + oldFile.getName());
            } else if (plugin.getConfigManager().isDebug()) {
                plugin.getLogger().info("Deleted legacy schematic: " + oldFile.getName());
            }
        }
    }
}

File: src/main/kotlin/com/regionrental/managers/WorldEditManager.kt
package com.regionrental.managers

import com.regionrental.RegionRental
import com.sk89q.worldedit.EditSession
import com.sk89q.worldedit.WorldEdit
import com.sk89q.worldedit.bukkit.BukkitAdapter
import com.sk89q.worldedit.extent.clipboard.BlockArrayClipboard
import com.sk89q.worldedit.extent.clipboard.Clipboard
import com.sk89q.worldedit.extent.clipboard.io.BuiltInClipboardFormat
import com.sk89q.worldedit.extent.clipboard.io.ClipboardFormats
import com.sk89q.worldedit.function.operation.ForwardExtentCopy
import com.sk89q.worldedit.function.operation.Operations
import com.sk89q.worldedit.regions.CuboidRegion
import com.sk89q.worldedit.session.ClipboardHolder
import com.sk89q.worldguard.WorldGuard
import org.bukkit.Bukkit
import org.bukkit.World
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap
import java.util.logging.Level

/**
 * Manages WorldEdit integration for block restoration.
 * Captures region state when rental starts and restores it when rental expires.
 * Uses Sponge Schematic format (.schem) for reliable cross-session persistence.
 */
class WorldEditManager(private val plugin: RegionRental) {

    private val schematicsFolder = File(plugin.dataFolder, "schematics")

    // LRU cache for clipboards - limits memory usage
    private val maxCacheSize: Int
    private val clipboardCache: MutableMap<String, Clipboard>

    // Track which schematics exist on disk (for hasCapture checks without loading)
    private val knownSchematics: MutableSet<String> = ConcurrentHashMap.newKeySet()

    companion object {
        private const val DEFAULT_CACHE_SIZE = 20
        private const val SCHEMATIC_EXTENSION = ".schem"
        private const val OLD_EXTENSION = ".dat"
    }

    init {
        maxCacheSize = plugin.configManager.getSchematicCacheSize(DEFAULT_CACHE_SIZE)

        // Create LRU cache with access-order ordering, wrapped for thread safety
        clipboardCache = Collections.synchronizedMap(object : LinkedHashMap<String, Clipboard>(maxCacheSize, 0.75f, true) {
            override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Clipboard>): Boolean {
                val shouldRemove = size > maxCacheSize
                if (shouldRemove && plugin.configManager.isDebug) {
                    plugin.logger.info("Evicting schematic from cache: ${eldest.key}")
                }
                return shouldRemove
            }
        })

        // Create schematics folder if it doesn't exist
        if (!schematicsFolder.exists()) {
            schematicsFolder.mkdirs()
        }

        // Index existing schematics (don't load them into memory)
        indexSchematics()
    }

    /**
     * Captures the current state of a WorldGuard region.
     */
    fun captureRegion(regionName: String, world: World): Boolean {
        if (!plugin.configManager.isBlockRestoration) {
            return false
        }

        return try {
            val wgRegion = getWorldGuardRegion(regionName, world)
            if (wgRegion == null) {
                plugin.logger.warning("Could not find WorldGuard region: $regionName in world ${world.name}")
                return false
            }

            val weWorld = BukkitAdapter.adapt(world)
            val min = wgRegion.minimumPoint
            val max = wgRegion.maximumPoint
            val region = CuboidRegion(weWorld, min, max)

            val clipboard = BlockArrayClipboard(region)

            WorldEdit.getInstance().newEditSession(weWorld).use { editSession ->
                val copy = ForwardExtentCopy(editSession, region, clipboard, region.minimumPoint)
                copy.isCopyingEntities = true
                copy.isCopyingBiomes = false
                Operations.complete(copy)
            }

            val compositeKey = "${world.name}:$regionName"
            clipboardCache[compositeKey] = clipboard
            knownSchematics.add(compositeKey)

            saveSchematic(compositeKey, clipboard)

            if (plugin.configManager.isDebug) {
                plugin.logger.info("Captured region state for: $regionName in world ${world.name}")
            }

            true
        } catch (e: Exception) {
            plugin.logger.log(Level.SEVERE, "Failed to capture region: $regionName in world ${world.name}", e)
            false
        }
    }

    @Deprecated("Use world-aware version", ReplaceWith("captureRegion(regionName, world)"))
    fun captureRegion(regionName: String): Boolean {
        val world = Bukkit.getWorlds()[0]
        return captureRegion(regionName, world)
    }

    /**
     * Restores a region to its captured state.
     */
    fun restoreRegion(regionName: String, world: World): Boolean {
        if (!plugin.configManager.isBlockRestoration) {
            return false
        }

        val compositeKey = "${world.name}:$regionName"
        val clipboard = getOrLoadClipboard(compositeKey)

        if (clipboard == null) {
            plugin.logger.warning("No saved state found for region: $regionName in world ${world.name}")
            return false
        }

        return try {
            val weWorld = BukkitAdapter.adapt(world)

            WorldEdit.getInstance().newEditSession(weWorld).use { editSession ->
                val holder = ClipboardHolder(clipboard)
                val operation = holder.createPaste(editSession)
                    .to(clipboard.origin)
                    .ignoreAirBlocks(false)
                    .build()
                Operations.complete(operation)
            }

            if (plugin.configManager.isDebug) {
                plugin.logger.info("Restored region state for: $regionName in world ${world.name}")
            }

            if (plugin.configManager.isAutoDeleteSchematics) {
                clipboardCache.remove(compositeKey)
                knownSchematics.remove(compositeKey)
                deleteSchematic(compositeKey)
            }

            true
        } catch (e: Exception) {
            plugin.logger.log(Level.SEVERE, "Failed to restore region: $regionName in world ${world.name}", e)
            false
        }
    }

    @Deprecated("Use world-aware version", ReplaceWith("restoreRegion(regionName, world)"))
    fun restoreRegion(regionName: String): Boolean {
        val world = Bukkit.getWorlds()[0]
        return restoreRegion(regionName, world)
    }

    /**
     * Checks if a region state has been captured (checks index, doesn't load from disk).
     */
    fun hasCapture(regionName: String): Boolean = regionName in knownSchematics

    /**
     * Gets the captured clipboard for a region (lazy loads from disk if needed).
     */
    fun getClipboard(regionName: String): Clipboard? = getOrLoadClipboard(regionName)

    /**
     * Deletes a captured region state from cache, index, and disk.
     */
    fun deleteCapture(regionName: String) {
        clipboardCache.remove(regionName)
        knownSchematics.remove(regionName)
        deleteSchematic(regionName)
    }

    /**
     * Gets the current cache size (for monitoring/debugging).
     */
    val cacheSize: Int
        get() = clipboardCache.size

    /**
     * Gets the total number of known schematics on disk.
     */
    val totalSchematicCount: Int
        get() = knownSchematics.size

    /**
     * Clears the memory cache (schematics remain on disk).
     */
    fun clearCache() {
        clipboardCache.clear()
        if (plugin.configManager.isDebug) {
            plugin.logger.info("Schematic cache cleared")
        }
    }

    private fun getOrLoadClipboard(compositeKey: String): Clipboard? {
        // Check cache first
        clipboardCache[compositeKey]?.let { return it }

        // Not in cache - check if it exists on disk
        if (compositeKey !in knownSchematics) {
            return null
        }

        // Load from disk
        val loaded = loadSchematic(compositeKey)
        if (loaded != null) {
            clipboardCache[compositeKey] = loaded
            if (plugin.configManager.isDebug) {
                plugin.logger.info("Lazy loaded schematic from disk: $compositeKey")
            }
        }
        return loaded
    }

    private fun getWorldGuardRegion(regionName: String, world: World) =
        WorldGuard.getInstance().platform.regionContainer
            .get(BukkitAdapter.adapt(world))
            ?.getRegion(regionName)

    private fun saveSchematic(regionName: String, clipboard: Clipboard) {
        val file = File(schematicsFolder, "$regionName$SCHEMATIC_EXTENSION")

        runCatching {
            FileOutputStream(file).use { fos ->
                BuiltInClipboardFormat.SPONGE_SCHEMATIC.getWriter(fos).use { writer ->
                    writer.write(clipboard)
                }
            }
            if (plugin.configManager.isDebug) {
                plugin.logger.info("Saved schematic: $regionName$SCHEMATIC_EXTENSION")
            }
        }.onFailure { e ->
            plugin.logger.log(Level.WARNING, "Failed to save schematic for: $regionName", e)
        }
    }

    private fun indexSchematics() {
        if (!schematicsFolder.exists()) return

        // Index new .schem files
        schematicsFolder.listFiles { _, name -> name.endsWith(SCHEMATIC_EXTENSION) }?.forEach { file ->
            val regionName = file.name.removeSuffix(SCHEMATIC_EXTENSION)
            knownSchematics.add(regionName)
        }

        // Migration: Handle legacy .dat files
        val datFiles = schematicsFolder.listFiles { _, name -> name.endsWith(OLD_EXTENSION) }
        if (datFiles != null && datFiles.isNotEmpty()) {
            plugin.logger.warning("Found ${datFiles.size} legacy .dat file(s) - these are EMPTY due to serialization bug in v2.5.1")
            if (plugin.configManager.isAutoDeleteSchematics) {
                var deleted = 0
                for (datFile in datFiles) {
                    if (datFile.delete()) deleted++
                }
                plugin.logger.info("Deleted $deleted empty legacy .dat file(s)")
                if (deleted < datFiles.size) {
                    plugin.logger.warning("Could not delete ${datFiles.size - deleted} legacy file(s) - check file permissions")
                }
            } else {
                plugin.logger.info("Set 'restoration.auto-delete-schematics: true' to auto-clean legacy files")
            }
        }

        if (knownSchematics.isNotEmpty()) {
            plugin.logger.info("Indexed ${knownSchematics.size} region schematic(s) (lazy loading enabled)")
        }
    }

    private fun loadSchematic(regionName: String): Clipboard? {
        val schemFile = File(schematicsFolder, "$regionName$SCHEMATIC_EXTENSION")

        if (schemFile.exists()) {
            return runCatching {
                val format = ClipboardFormats.findByFile(schemFile)
                if (format == null) {
                    plugin.logger.warning("Unknown schematic format: ${schemFile.name}")
                    return null
                }

                FileInputStream(schemFile).use { fis ->
                    format.getReader(fis).use { reader ->
                        val clipboard = reader.read()
                        if (plugin.configManager.isDebug) {
                            plugin.logger.info("Loaded schematic: $regionName$SCHEMATIC_EXTENSION")
                        }
                        clipboard
                    }
                }
            }.onFailure { e ->
                plugin.logger.log(Level.WARNING, "Failed to load schematic: ${schemFile.name}", e)
            }.getOrNull()
        }

        // Migration: Warn about legacy .dat files
        val oldFile = File(schematicsFolder, "$regionName$OLD_EXTENSION")
        if (oldFile.exists()) {
            plugin.logger.warning("Legacy .dat schematic for '$regionName' is empty (serialization bug). Will be re-captured on next rental.")
            if (plugin.configManager.isAutoDeleteSchematics) {
                if (!oldFile.delete()) {
                    plugin.logger.warning("Could not delete legacy schematic: ${oldFile.name}")
                }
            }
        }

        return null
    }

    private fun deleteSchematic(regionName: String) {
        // Delete new .schem file
        val schemFile = File(schematicsFolder, "$regionName$SCHEMATIC_EXTENSION")
        if (schemFile.exists()) {
            if (!schemFile.delete()) {
                plugin.logger.warning("Could not delete schematic: ${schemFile.name}")
            } else if (plugin.configManager.isDebug) {
                plugin.logger.info("Deleted schematic: ${schemFile.name}")
            }
        }

        // Also delete legacy .dat file if it exists
        val oldFile = File(schematicsFolder, "$regionName$OLD_EXTENSION")
        if (oldFile.exists()) {
            if (!oldFile.delete()) {
                plugin.logger.warning("Could not delete legacy schematic: ${oldFile.name}")
            } else if (plugin.configManager.isDebug) {
                plugin.logger.info("Deleted legacy schematic: ${oldFile.name}")
            }
        }
    }
}
